src/compute-client/src/controller/instance.rs

```diff
@@ -227,9 +261,14 @@ where
 
     /// Marks the end of any initialization commands.
     ///
-    /// Intended to be called by `Controller`, rather than by other code (to avoid repeated calls).
+    /// Intended to be called by `Controller`, rather than by other code.
+    /// Calling this method repeatedly has no effect.
     pub fn initialization_complete(&mut self) {
-        self.send(ComputeCommand::InitializationComplete);
+        // The compute protocol requires that `InitializationComplete` is sent only once.
+        if !self.initialized {
+            self.send(ComputeCommand::InitializationComplete);
+            self.initialized = true;
```

Comment:
What will happen if the process dies is restarted between these two lines:
```rust
self.send(ComputeCommand::InitializationComplete);
self.initialized = true;
```
Are we going to get two copies of `ComputeCommand::InitializationComplete` in violation of the compute control requirements from the comment above?

Reply:
I think the order is correct / doesn't matter here. If `envd` dies between those lines, we will re-send `InitializationComplete` in the next protocol incarnation, but that will happen anyway because we don't persist `initialized`.

Reply:
That's right. If environmentd restarts it also needs to reconnect to the replicas, which causes a new protocol iteration, which starts in the initialization stage again.